### DISCUSSION TITLE
What is ethereum used for in crossbar?

### DISCUSSION THREAD

**Discussion by sethrh** [crossbario/crossbar#2062] [Q&A] [answered]
I've just run across the bug here https://github.com/crossbario/crossbar/pull/2059, which alerted me to there being ethereum / block chain code inside crossbar and autobahn. I searched around for some documentation on this and couldn't find any. 

I can't think of any way that ethereum would be related to WAMP (my first reaction was that someone snuck a miner into the code somehow). 

Can someone enlighten me?

**Comment by oberstet** [comment]
> I can't think of any way that ethereum would be related to WAMP (my first reaction was that someone snuck a miner into the code somehow).

sure! here is an example. imagine

* Alice wants to own (== control) a WAMP routing realm for her application, and
* Bob and Joe each know how to host and operate WAMP router nodes

all 3 are independent actors. to establish trust relations for above scenario, they need a common trust anchor. and to do that without 2 of them relying on the 3rd or any other centralized party they can use a programmable blockchain: Ethereum

the code in autobahn/crossbar already goes a long way, but is still experimental, as there are yet unsolved questions, the most important one:

with Ethereum integration, Alice, Bob and Joe _can_ establish trust relations in non-centralized fashion, but _why_ would they want to?

Alice has the need for a routing realm, because she supposedly has a need for her application.

But why would Bob and Joe have a motivation to provide their routing services to Alice's realm?

IOW: besides the trust relations, what about economic relations?

**Reply by sethrh** [reply]
Thanks. I think I understand. 

If multiple sysadmins want to run the same application that is able to cooperate and share information, but none of the sysadmins trust each other. 

Maybe something like Mastodon, many admins running a federated application. A user on Alice's application publishes information to a topic, a user on Bob's application wants to see that information. Bob's application is able to know that information on topics coming from Alice's application are valid by consulting the shared block chain, even though Bob may not know/trust Alice directly. In other words, all of the admins of that application wouldn't need to share credentials for their applications to share data. 

How is performance? Does each publish get an entry into the block chain, or is it only used when connecting routers together?

**Comment by oberstet** [comment, marked as answer]
Yes, you are on the right track! Mastodon or generally federated systems are a good category of systems to compare this experimental feature to. However, it goes quite a bit further.

To begin with, there is a distinction made between "operationally dependent on" and "trust wrt to some function". And continuing my example with Alice, Joe and Bob:

* Alice doesn't want to operationally depend on only Joe or only Bob - she want users of her realm or her application to have her realm available, and be able to connect to "some" endpoint for WAMP routing in that realm
* users/apps of Alice's realm want to be able to connect to "some" endpoint, but only trust Alice for the definition of the realm, and not operationally depend on only Joe or only Bob
* Joe wants to host Alice's realm (because of unsolved economic incentives). and to provide 1 shared routing for the 1 realm, Joe needs to connect his routing node to Bob's - without operationally become dependent on Bob

and so on, I guess you get the idea.

Next, rgd "performance": all of above is essentially providing information to configure WAMP routing nodes, as eg Crossbar.io contains in its node key and node config file - but in a decentralized database (Ethereum). Thus, there is no performance impact at the messaging level.

Finally, the experimental code goes beyond in yet another way: it provides end-to-end encrypted application traffic. That is, WAMP client connecting to Alice's realm communicate in a privacy preserving manner without even Bob or Joe able to read. This feature is a mini-world to explain .. so unless you really want to know .. I'm skipping this for now;)

**Comment by oberstet** [comment]
So for completeness, here is the list of actors in my example:

1. Alice: realm owner
2. Joe: a first routing hoster
3. Bob: a second routing hoster
4. Carol: a first user of Alice's realm
5. Homer: a second user of Alice's realm


Alice wants to control her realm, but doesn't want to host routing services, and doesn't want to depend on a centralized party (only on Bob or only on Joe).

Carol and Homer trust Alice, but do not want to operationally depend on Bob only or on Joe only.

Carol and Homer want to use WAMP routing to have their apps in Alice's realm talk to each other, but they do not want anyone other to read or modify that app traffic (this includes the WAMP router nodes!).

And so on ..
